instructionAPI/src/MultiRegister.C

```diff
@@ -0,0 +1,151 @@
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include "MultiRegister.h"
+#include <vector>
+#include <set>
+#include <sstream>
+#include "Visitor.h"
+#include "../../common/src/singleton_object_pool.h"
+#include "InstructionDecoder-power.h"
+#include "registers/MachRegister.h"
+#include "Architecture.h"
+#include "ArchSpecificFormatters.h"
+#include "../../common/h/compiler_diagnostics.h"
+#include "registers/x86_regs.h"
+
+using namespace std;
+
+namespace Dyninst
+{
+  namespace InstructionAPI
+  {
+    MultiRegisterAST::MultiRegisterAST(MachRegister r, uint32_t num_elements) :
+            Expression(r,num_elements), consecutive(true)
+    {
+        uint32_t regVal = r.val();
+        for (uint32_t i = 0; i < num_elements; i++){
+            m_Regs.push_back(make_shared(singleton_object_pool<RegisterAST>::construct(MachRegister(regVal+i), 0, r.size() * 8,num_elements)));
+        }
+    }
+
+    MultiRegisterAST::MultiRegisterAST(std::vector<RegisterAST::Ptr> inputRegASTs) : Expression(inputRegASTs[0]->getID(),inputRegASTs.size()), m_Regs{std::move(inputRegASTs)}
+    {
+    }
+
+    void MultiRegisterAST::getChildren(vector<InstructionAST::Ptr>& /*children*/) const
+    {
+        return;
+    }
+    void MultiRegisterAST::getChildren(vector<Expression::Ptr>& /*children*/) const
+    {
+        return;
+    }
+    void MultiRegisterAST::getUses(set<InstructionAST::Ptr>& uses)
+    {
+        for (const auto & m_Reg : m_Regs) {
+            m_Reg->getUses(uses);
+        }
+    }
+    bool MultiRegisterAST::isUsed(InstructionAST::Ptr findMe) const
+    {
+        for (const auto &regAST : m_Regs) {
+            if (findMe->checkRegID(regAST->getID(), regAST->lowBit(), regAST->highBit()))
+                return true;
+        }
+        return false;
+    }
+
+    std::string MultiRegisterAST::format(Architecture arch, formatStyle) const
+    {
+        if(arch == Arch_amdgpu_gfx908 || arch == Arch_amdgpu_gfx90a || arch == Arch_amdgpu_gfx940){
+            return AmdgpuFormatter::formatMultiRegister(m_Regs[0]->getID(),m_Regs.size());
+        }
+        assert(0 && " multi register currently defined for amdgpu formats only ");
+        return std::string("");
+    }
+
+    std::string MultiRegisterAST::format(formatStyle) const
+    {
+        std::string ret("[");
+        for (auto & m_Reg : m_Regs) {
+            ret.append(m_Reg->format());
+            ret.append(",");
+        }
+        auto i = ret.rfind(",");
+        ret.erase(i,1);
```

Comment:
More efficient to only add the ',' if needed instead of finding and deleting it:

```suggestion
        bool firstEntry{true};
        std::string ret("[");
        for (auto & m_Reg : m_Regs) {
            if (!firstEntry)  {
                firstEntry = false;
                ret.append(",");
            }
            ret.append(m_Reg->format());            
        }
```